### DISCUSSION TITLE
API rule using a @request Relation that can have multiple values

### DISCUSSION THREAD

**Discussion by JossLadanyi** [pocketbase/pocketbase#1808] [Q&A] [answered]
Hi,

First of all: wonderful job on Pocketbase, thank you for the great tool!
Sorry for the newbie question, I am quite new to Pocketbase.
I have an owned and rented field for the users table. Both are relation types, referring to a second table: books. Owned can have multiple values (Max Select empty), while rented can have only one value (Max Select = 1).
I would like to set an API rule for the users table so that users that own a book could list users that rented that book (and so expand my query with those users when listing books).
I am trying
@request.auth.owned ~ rented
but that does not work. Where did I go wrong?
Also: is it possible to get an API rule if rented can also have multiple values?
Would all be easier if books had the renter and owner fields instead (both with Max Select = 1)?
Thank you so much!

Regards,

Joseph

**Comment by ganigeorgiev** [comment, marked as answer]
Are you sure that you are authenticated? I've just tested the above rule and it is working fine for me.

The only reason I can think of the rule may not work is if `rented` initially was saved as multiple `relation` field (aka. "Max select != 1") but later changed to a single. This is because we are not updating the previously stored field values on constraint changes since it could have some unexpected results and the original value is preserved allowing developers to specify [migration](https://pocketbase.io/docs/migrations/) and normalize the value based on their use case (eg. if it had 10 values, keep only the last one).

`~` operator is translated to a SQL `LIKE` and internally the above rule compares the serialized json literal against the rented column value, something like:
```sql
'["id1", "id2", "id3"]' LIKE '%' || RENTED || '%'
```

A more universal solution that should work even if `rented` is multi-`relation` field would be to narrow the rel fields using dot-notation and use `?=` operator ("at least one of equal"):

```js
// ensures that the current authenticated user has owned books to prevent showing users that don't have rented books
// AND
// at least one of the current authenticated user "owned" books id must be equal to at least one of the rented books id of the other user
@request.auth.owned:length > 0 && @request.auth.owned.id ?= rented.id
```

_**update**: added check to ensure that the current auth user has owned books_

**Reply by JossLadanyi** [reply]
Wow, **you are incredible!** Thank you for that swift reply!
Indeed it might have been the case with Max select being set to a different value and reset to 1 later... You alternate suggestion works like a charm!
I set it for both the list and view action. Now there's only one thing I do not understand: after this I should be able to query all books owned by the user (this part works) and expand it with the renters like: (javascript)
```
pb.collection('books').getFullList(200, {
	filter: `"${locals.user.owned}" ~ id`,
	expand: 'users(rented)'
})
```
And yet, the expand blocks of the result remain empty. Shouldn't that work? Do you have any suggestions?

**Reply by ganigeorgiev** [reply]
Indirect expansion currently works only with a single `relation` fields (_this and other limitations could be found at the bottom of https://pocketbase.io/docs/expanding-relations/#indirect-expand; support for indirect multi-`relation` field expansion is in the roadmap, but there could be some performance overhead and I have to test it more thoroughly before enabling it_) and it is possible that if previously the `rented` field was with `Max select != 1` to fail the expansion (because the stored value in the database will be a serialized json array and not just the rented book id).

You can try to resave the user(s) to normalize the value back to a single one.
Eventually, I'll consider applying the multi<->single value normalization automatically since this seems to be a common issue (eg. https://github.com/pocketbase/pocketbase/discussions/1545).

**Reply by JossLadanyi** [reply]
For the record: yes, that was it, resaving users solved the issue, expand is filled as expected.
Thank you again for your exceptional help!!!
